src/triples.rs

```diff
@@ -265,6 +269,119 @@ impl<'de> serde::Deserialize<'de> for TriplesBitmap {
+        let mut cv = CompactVector::with_capacity(entries, sucds::utils::needed_bits(entries))
+            .expect("Failed to create OPS index compact vector.");
+        // disable parallelization temporarily for easier debugging
+        //let wavelet_y = wavelet_thread.join().unwrap(); // join as late as possible for max parallelization
+        for mut indices in indicess {
+            let mut first = true;
+            // sort by predicate
+            indices.sort_by_cached_key(|pos_y| wavelet_y.access(*pos_y as usize).unwrap());
+            for index in indices {
+                bitmap_index_bitvector.push_bit(first);
+                first = false;
+                cv.push_int(index as usize).unwrap();
+            }
+        }
+        let bitmap_index = Bitmap { dict: Rank9Sel::new(bitmap_index_bitvector) };
+        let op_index = OpIndex { sequence: cv, bitmap: bitmap_index };
+        debug!("built OPS index");
+        Self { order, bitmap_y, adjlist_z, op_index, wavelet_y }
+    }
+
+    /// Creates a new TriplesBitmap from a list of sorted RDF triples
+    // under construction, get it to compile with empty or nonsense values first, then fix
+    // originally written by Greg Hanson in the experimental-rdf2hdt PR, move here is in progress
+    pub fn from_triples(triples: Vec<TripleId>) -> Self {
+        let mut y_bitmap = BitVector::new();
+        let mut z_bitmap = BitVector::new();
+        let mut array_y = Vec::new();
+        let mut array_z = Vec::new();
+
+        let mut last_x = 0;
+        let mut last_y = 0;
+        let mut last_z = 0;
+
+        for (i, triple) in triples.iter().enumerate() {
+            let x = triple.subject_id;
+            let y = triple.predicate_id;
+            let z = triple.object_id;
+
+            assert!(!(x == 0 || y == 0 || z == 0), "triple IDs should never be zero");
+
+            if i == 0 {
+                array_y.push(y);
+                array_z.push(z);
+            } else if x != last_x {
+                assert!(!(x != last_x + 1), "the subjects must be correlative.");
+
+                //x unchanged
+                y_bitmap.push_bit(true);
+                array_y.push(y);
+
+                z_bitmap.push_bit(true);
+                array_z.push(z);
+            } else if y != last_y {
+                assert!(!(y < last_y), "the predicates must be in increasing order.");
+
+                // y unchanged
+                y_bitmap.push_bit(false);
+                array_y.push(y);
+
+                z_bitmap.push_bit(true);
+                array_z.push(z);
+            } else {
+                assert!(!(z < last_z), "the objects must be in increasing order");
+
+                // z changed
+                z_bitmap.push_bit(false);
+                array_z.push(z);
+            }
+
+            last_x = x;
+            last_y = y;
+            last_z = z;
+        }
+        y_bitmap.push_bit(true);
```

Comment:
https://github.com/DeciSym/rdf2hdt/blob/main/src/bitmap_triples.rs#L96

Argh this was in the original implementation, I don't know how we both missed this during the port. Pretty sure the trailing set bit is described as part of the HDT spec